subsys: usb: add loopback function

jfischer-no opened the pull request:
Add loopback function. This function can be used to test
USB device drivers and device stack connected to linux host
and has the similar interface as "Gadget Zero" [1] of the Linux
kernel.

Use modprobe usbtest to load the module, see also [2] for the
description of the tests and for Vendor and Product ID of the
"Gadget Zero". The userspace tool testusb [3] is needed to start
the tests.

[1] linux/drivers/usb/gadget/function/f_loopback.c
[2] linux/drivers/usb/misc/usbtest.c
[3] linux/tools/usb/testusb.c

jfischer-no committed:
usb: add callback codes for Set/Clear Feature ENDPOINT_HALT

Add callback codes for Set/Clear Feature ENDPOINT_HALT.
These can be used to inform a function that the device stack has
received Set/Clear Feature request ENDPOINT_HALT. The function can
then abort or restart the transfer accordingly.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
subsys: usb: stall if there is no data buffer

Return a STALL in the setup or data stage if there is no data buffer.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

loicpoulain reviewed (commented):
Note that according to USB specification, a function cannot issue a STALL packet in response to a a SETUP packet. A function must ACK the SETUP or ignore it (cf Control Transfers chapter).

jfischer-phytec-iot commented:
@loicpoulain In what context do you mean it?

loicpoulain commented:
https://github.com/zephyrproject-rtos/zephyr/pull/7435/commits/bcd517953919e34838482fd964d920e572ec9d5d

jfischer-phytec-iot commented:
@loicpoulain This protocol STALL is valid, e.g. if the callback and buffer for vendor request are not proper initialized, then the stack should return STALL (in response to any IN or OUT transaction, see 8.5.3.4) because the device is unable to complete the command.

loicpoulain commented:
Yes you're right, this looks good indeed because the STALL will not be sent in response to the SETUP transaction/stage but to the following one(s) (IN or OUT)...

carlescufi commented:
CC @pawelzadrozniak you can use this to test your driver

jfischer-no committed:
subsys: usb: class: add loopback function

Add loopback function. This function can be used to test
USB device drivers and device stack connected to linux host
and has the similar interface as "Gadget Zero" [1] of the Linux
kernel.

Use modprobe usbtest to load the module, see also [2] for the
description of the tests and for Vendor and Product ID of the
"Gadget Zero". The userspace tool testusb [3] is needed to start
the tests.

[1] linux/drivers/usb/gadget/function/f_loopback.c
[2] linux/drivers/usb/misc/usbtest.c
[3] linux/tools/usb/testusb.c

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

finikorg commented on subsys/usb/class/loopback.c:
```diff
@@ -0,0 +1,156 @@
+
+	usb_read(ep, NULL, 0, &bytes_to_read);
+	SYS_LOG_DBG("ep 0x%x, bytes to read %d ", ep, bytes_to_read);
+	usb_read(ep, loopback_buf, bytes_to_read, NULL);
```
should it send data back to IN?

jfischer-phytec-iot replied:
IN endpoint (buffer) is always prepared for the IN token and independent of OUT endpoint.

finikorg reviewed (changes requested):
I would add parts from commit message to documentation (in a follow up commit). We should also mention how to attach usbtest driver to this USB device.
BTW: Are you able to pass tests? For quark_se most of tests are failing.

jfischer-phytec-iot commented on subsys/usb/usb_descriptor.c:
```diff
@@ -760,6 +767,46 @@ static struct dev_common_descriptor common_desc = {
+			.bDescriptorType = USB_INTERFACE_DESC,
+			.bInterfaceNumber = FIRST_IFACE_LOOPBACK,
+			.bAlternateSetting = 0,
+			.bNumEndpoints = 2,
```
there are always two, directly below :-)

finikorg replied:
The point was to omit "magic numbers" when possible.

finikorg commented:
@jfischer-phytec-iot Thanks for docs, my mistake was that I changed vendor/product pair in menuconfig (and with wrong device requiring to change alt setting, etc). It would be good to add this info to Zephyr docs or Readme.

carlescufi commented:
@jfischer-phytec-iot can you please rebase this? Aside from that it's ready to merge

galak commented:
Should I merge this?

carlescufi commented:
@galak yes please, this one flew under my radar
